src/plugins/internalnotification/internalnotification.c

```diff
@@ -0,0 +1,312 @@
+/**
+ * @file
+ *
+ * @brief Source for internalnotification plugin
+ *
+ * @copyright BSD License (see doc/COPYING or http://www.libelektra.org)
+ *
+ */
+
+#include "internalnotification.h"
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include <kdbassert.h>
+#include <kdbhelper.h>
+#include <kdblogger.h>
+
+/**
+ * Structure for registered key variable pairs
+ * @internal
+ */
+struct _KeyRegistration
+{
+	char * name;
+	int * variable;
+	struct _KeyRegistration * next;
+};
+typedef struct _KeyRegistration KeyRegistration;
+
+/**
+ * Structure for internal plugin state
+ * @internal
+ */
+struct _PluginState
+{
+	KeyRegistration * head;
+	KeyRegistration * last;
+};
+typedef struct _PluginState PluginState;
+
+/**
+ * Creates a new KeyRegistration structure and appends it at the end of the registration list
+ * @internal
+ *
+ * @param pluginState		internal plugin data structure
+ *
+ * @return pointer to created KeyRegistration structure or NULL if memory allocation failed
+ */
+static KeyRegistration * elektraInternalnotificationAddNewRegistration (PluginState * pluginState)
+{
+	KeyRegistration * item = elektraMalloc (sizeof *item);
+	if (item == NULL)
+	{
+		return NULL;
+	}
+	item->next = NULL;
+
+	if (pluginState->head == NULL)
+	{
+		// Initialize list
+		pluginState->head = pluginState->last = item;
+	}
+	else
+	{
+		// Make new item end of list
+		pluginState->last->next = item;
+		pluginState->last = item;
+	}
+
+	return item;
+}
+
+/**
+ * Updates all KeyRegistrations according to data from the given KeySet
+ * @internal
+ *
+ * @param pluginState		internal plugin data structure
+ * @param keySet 				key set retrieved from hooks
+ *                   		(e.g. elektraInternalnotificationGet or elektraInternalnotificationSet)
+ *
+ */
+static void elektraInternalnotificationUpdateRegisteredKeys (PluginState * pluginState, KeySet * keySet)
+{
+	KeyRegistration * registeredKey = pluginState->head;
+	while (registeredKey != NULL)
+	{
+		Key * key = ksLookupByName (keySet, registeredKey->name, 0);
+		if (key != 0)
+		{
+			ELEKTRA_LOG ("found registeredKey=%s; updating variable=%p with string value \"%s\"", registeredKey->name,
+				     (void *)registeredKey->variable, keyString (key));
+
+			// Convert string value to long
+			char * end;
+			errno = 0;
+			long int value = strtol (keyString (key), &end, 10);
+			// Update variable if conversion was successful and did not exceed integer range
+			if (*end == 0 && errno == 0 && value <= INT_MAX && value >= INT_MIN)
+			{
+				*(registeredKey->variable) = value;
+			}
+			else
+			{
+				ELEKTRA_LOG ("conversion failed! keyString=\"%s\" *end=%c, errno=%d, value=%ld", keyString (key), *end,
```

Comment:
Some higher log level here (or even better, add an warning), the user did something wrong here.